Public/Src/Pips/Dll/Operations/Process.cs

```diff
@@ -765,6 +793,119 @@ public bool TryComputePipUniqueOutputHash(PathTable pathTable, out long pipUniqu
+            PathTable pathTable,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputFiles = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputDirectroies = null)
+        {
+            // sourceChangeAffectedOutputFiles and sourceChangeAffectedOutputDirectroies 
+            // are initilized with the source change list from GBR.
+            // If it is null, that means no file or directory change for this build.
+            // Thus no output is affected by the change.
+            if (!sourceChangeAffectedOutputFiles.Any() && !sourceChangeAffectedOutputDirectroies.Any())
+            {
+                return false;
+            }
+
+            // Check if any dynamic and static file dependency is sourceChangeAffectedOutputFiles
+            var hasAffected = sourceChangeAffectedOutputFiles.Intersect(dynamicallyObservedFiles).Any()
+                || sourceChangeAffectedOutputFiles.Intersect(Dependencies.Select(d => d.Path)).Any();
+
+            if (hasAffected)
+            {
+                return true;
+            }
+
+            if (dynamicallyObservedEnumerations.Any() || DirectoryDependencies.Any())
+            {
+                // Check if dynamic and static directory denpendencis under any directory of sourceChangeAffectedOutputDirectroies
+                foreach (var affectedDir in sourceChangeAffectedOutputDirectroies)
+                {
+                    foreach (var dynamicDir in dynamicallyObservedEnumerations)
+                    {
+                        if (affectedDir.TryGetRelative(pathTable, dynamicDir, out var relativePath))
+                        {
+                            return true;
+                        }
+                    }
+
+                    foreach (var staticDir in DirectoryDependencies)
+                    {
+                        if (affectedDir.TryGetRelative(pathTable, staticDir.Path, out var relativePath))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                // check if the dynamic or static directroy dependency contain any file from sourceChangeAffectedOutputFiles
+                foreach (var affectedFile in sourceChangeAffectedOutputFiles)
+                {
+                    foreach (var dynamicDir in dynamicallyObservedEnumerations)
+                    {
+                        if (dynamicDir.TryGetRelative(pathTable, affectedFile, out var relativePath))
+                        {
+                            return true;
+                        }
+                    }
+
+                    foreach (var staticDir in DirectoryDependencies)
+                    {
+                        if (staticDir.Path.TryGetRelative(pathTable, affectedFile, out var relativePath))
+                        {
+                            return true;
+                        }
+                    }

```

Comment:
We cannot statically check if a path is relative to another. We need to consult to `FileContentManager`. Thus, I don't think this method belongs here.

For example, suppose that your pip depends on a seal directory `D` with explicitly specified members `D\a` and `D\b`. If the affected input contains only `D\c`, then, with your method, this pip is affected, but it is actually not.